src/compile/invalid/ScaleInvalidDataMode.ts

```diff
@@ -0,0 +1,51 @@
+import {SignalRef} from 'vega';
+import {ScaleChannel} from '../../channel';
+import {Config} from '../../config';
+import {MarkInvalidDataMode} from '../../invalid';
+import {MarkDef, isPathMark} from '../../mark';
+import {ScaleType, hasContinuousDomain} from '../../scale';
+import {getMarkPropOrConfig} from '../common';
+import {normalizeInvalidDataMode} from './normalizeInvalidDataMode';
+
+export type ScaleInvalidDataMode =
+  // remove 'break-paths-and-keep-path-domains' from MarkInvalidDataMode
+  // because it is a macro for '"filter"' or `"break-path-keep-domains`
+  | Omit<MarkInvalidDataMode, 'break-paths-and-keep-path-domains'>
+
+  // Add always-valid because at scale level, categorical scales can handle any values and thus is always valid.
+  | 'always-valid';
+
+export function getScaleInvalidDataMode<C extends ScaleChannel>({
+  markDef,
+  config,
+  scaleChannel,
+  scaleType,
+  isCountAggregate
+}: {
+  markDef: MarkDef;
+  config: Config<SignalRef>;
+  scaleChannel: C;
+  scaleType: ScaleType;
+  isCountAggregate: boolean;
+}): ScaleInvalidDataMode {
+  if (!scaleType || !hasContinuousDomain(scaleType) || isCountAggregate) {
```

Comment:
What's `!scaleType` checking? If scaleType === undefined, it's linear?

Reply:
when there is no scale.

Reply:
Btw, one thing I hate about Vega-lite code is that it's written before Typescript has `strictNullChecks` and we haven't migrated to turn it on.

Reply:
In which case does the channel have no scale?